x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java

```diff
@@ -6,70 +6,102 @@
 package org.elasticsearch.xpack.sql.expression.function.scalar.datetime;
 
 import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.xpack.ql.expression.gen.processor.Processor;
 import org.elasticsearch.xpack.sql.SqlIllegalArgumentException;
 import org.elasticsearch.xpack.sql.util.DateUtils;
 
 import java.io.IOException;
 import java.time.DateTimeException;
 import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetTime;
 import java.time.ZoneId;
 import java.time.ZonedDateTime;
+import java.time.ZoneOffset;
 import java.time.format.DateTimeFormatter;
 import java.time.temporal.TemporalAccessor;
+import java.time.temporal.TemporalQuery;
 import java.util.Locale;
 import java.util.Objects;
+import java.util.function.BiFunction;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
 
 public class DateTimeParseProcessor extends BinaryDateTimeProcessor {
 
+    public enum Parser {
+        DATE_TIME("datetime", ZonedDateTime::from, LocalDateTime::from), 
+        TIME("time", OffsetTime::from, LocalTime::from);
+        
+        private final BiFunction<String, String, TemporalAccessor> parser;
+        
+        private final String parseType;
+
+        Parser(String parseType,  TemporalQuery<?>... queries) {
+            this.parseType = parseType;
+            this.parser = (timestampStr, pattern) -> DateTimeFormatter.ofPattern(pattern, Locale.ROOT)
+                    .parseBest(timestampStr, queries);
+        }
+
+        public Object parse(Object timestamp, Object pattern, ZoneId zoneId) {
+            if (timestamp == null || pattern == null) {
+                return null;
+            }
+            if (timestamp instanceof String == false) {
+                throw new SqlIllegalArgumentException("A string is required; received [{}]", timestamp);
+            }
+            if (pattern instanceof String == false) {
+                throw new SqlIllegalArgumentException("A string is required; received [{}]", pattern);
+            }
+
+            if (((String) timestamp).isEmpty() || ((String) pattern).isEmpty()) {
+                return null;
+            }
+            try {
+                TemporalAccessor ta = parser.apply((String) timestamp, (String) pattern);
+                if (ta instanceof LocalDateTime) {
+                    return DateUtils.atTimeZone((LocalDateTime) ta, zoneId);
+                } else if (ta instanceof LocalTime) {
+                    return OffsetTime.of((LocalTime) ta, ZoneOffset.UTC);
```

Comment:
The user session zoneId should be used.
An if for the `OffsetTime` should be added to apply the session zoneId also for times with a zone in the pattern.

Reply:
```java
Instant instant = Instant.now();
instant.atZone(zoneId);
((OffsetTime)ta).withOffsetSameInstant(zoneId.getRules().getStandardOffset(instant));
```
Is this the right way？